### PR TITLE
I'm Gonna Be (500 Lines) by The Containers - Migration of hello world app to three container layout.

### PR DESCRIPTION
**Pull request by tariqajyusuf**
🎵 When I commit, well I know I'm gonna change,
I'm gonna change the containers to three from two
When I push up, yeah I know it's gonna be
It's gonna be easier when we just `tilt up`
If the test breaks, well I know I'm gonna be
I'm gonna be the one who tries to get them fixed
And if I mess up, yeah I know it's gonna be
It's gonna be so hard to find a fix
But I would write 500 lines
And I would write 500 more
Just to make a change that's fourteen thousand lines
To improve so much more 🎵 

---

This is the big one. A full migration of the `helloworld` project to a full three-container layout. Upon approval this will be merged and then pushed to an NY-owned repo. There's a long list of changes that are coming here:

- Add `Tiltfiles` to more easily bring up and down the environment.
- Include @egracer's integration test changes
- Add an ingress controller to proxy traffic in the same way the existing app works.
- Spin off the frontend to it's own container running Node.js with Next.js.
- Multi-stage the builds more so that it doesn't take much longer when dependency libraries aren't changed.
- Fix redirects on the Spring Boot side to prevent having to call `/api/api`
and of course
- Migrate existing backend code
- Migrate frontend code and convert it to Next.js

I've checked that:

- [x] Tests pass
- [x] Appropriate changes to README are included in PR
- [x] Feature has been fleshed out in design document

**What isn't in this change?**

- Some changes to the GitHub Workflows that can't be done until I push it into a new repo. I'll do those separately.
- Coverage code changes for hello world.

### CODE REVIEW

File: helloworld/Tiltfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+k8s_yaml(kustomize('kubernetes'))
+
+docker_build(
+  'google-org/hello-world-redux-backend',
```
Comment (kebrahim) [inline helloworld/Tiltfile:4]: should this reference redux?

File: helloworld/Tiltfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  'backend',
+)
+
+docker_build(
+  'google-org/hello-world-redux-frontend',
```
Comment (kebrahim) [inline helloworld/Tiltfile:9]: should this reference redux?

File: helloworld/backend/build.gradle [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+dependencies {
+  implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+  implementation 'org.springframework.boot:spring-boot-starter-data-rest'
+  implementation 'org.springframework.boot:spring-boot-starter-data-mongodb'
+  runtimeOnly 'com.h2database:h2'
+  testImplementation('org.springframework.boot:spring-boot-starter-test') {
+    exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
+  }
+  testImplementation('org.junit.jupiter:junit-jupiter:5.4.0')
```
Comment (kebrahim) [inline helloworld/backend/build.gradle:23]: please also add 'testImplementation('org.junit.platform:junit-platform-launcher:1.5.2')' so we can run unit tests directly in VSCode

File: helloworld/README.md [hunk 4]
```diff
@@ -1,20 +1,65 @@
...
+
+### Start the app
+
+```
+tilt up
+```
+
+Then get the IP address for the app and connect to it on port 80:
+```
+kubectl describe ingress
```
Comment (kebrahim) [inline helloworld/README.md:64]: just confirming this is the same command you'd use on Mac

File: helloworld/backend/build.gradle [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+group = 'org.google'
+version = '0.0.1-SNAPSHOT'
+sourceCompatibility = '1.8'
+
+repositories {
+	mavenCentral()
+}
+
+dependencies {
```
Comment (kebrahim) [inline helloworld/backend/build.gradle:15]: did you purposefully take out the HAL browser dependency?

File: helloworld/kubernetes/backend.yaml [hunk 6]
```diff
@@ -1,51 +1,51 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: hw-web-service
+  name: hwr-backend-service
   labels:
-    app: hello-world
+    app: hello-world-redux
```
Comment (kebrahim) [inline helloworld/kubernetes/backend.yaml:6]: should this refer to "redux" (here & below)?

File: helloworld/kubernetes/frontend.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: hwr-frontend-service
+  labels:
+    app: hello-world-redux
```
Comment (kebrahim) [inline helloworld/kubernetes/frontend.yaml:6]: similar comment about "redux"?

File: helloworld/kubernetes/kustomization.yaml [hunk 8]
```diff
@@ -1,7 +1,9 @@
 secretGenerator:
-  - name: hw-mongo-pass
+  - name: hwr-mongo-pass
     literals:
       - password=ABADIDEA
```
Comment (kebrahim) [inline helloworld/kubernetes/kustomization.yaml:4]: should we leave a comment about how this should actually be handled?

File: helloworld/kubernetes/mongodb.yaml [hunk 9]
```diff
@@ -1,23 +1,23 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: hw-mongo-service
+  name: hwr-mongo-service
   labels:
-    app: hello-world
+    app: hello-world-redux
```
Comment (kebrahim) [inline helloworld/kubernetes/mongodb.yaml:6]: similar comment about "redux"

File: helloworld/Tiltfile (hunk 1 above)
Comment (tariqajyusuf) [reply helloworld/Tiltfile:4]: Ah thought I got all those, thanks for grabbing that one.

File: helloworld/backend/build.gradle (hunk 3 above)
Comment (tariqajyusuf) [reply helloworld/backend/build.gradle:23]: Done

File: helloworld/README.md (hunk 4 above)
Comment (tariqajyusuf) [reply helloworld/README.md:64]: Yes this is confirmed.

File: helloworld/kubernetes/backend.yaml (hunk 6 above)
Comment (tariqajyusuf) [reply helloworld/kubernetes/backend.yaml:6]: Yes, didn't quite get all of them

File: helloworld/kubernetes/kustomization.yaml (hunk 8 above)
Comment (tariqajyusuf) [reply helloworld/kubernetes/kustomization.yaml:4]: There is one in the README.md, I'll add a backlog bug in the starter repo to track this.

File: helloworld/backend/build.gradle (hunk 5 above)
Comment (kebrahim) [reply helloworld/backend/build.gradle:15]: what was the resolution here?

Comment (tariqajyusuf) [reply helloworld/backend/build.gradle:15]: Whoops looks like my comment didn't make it in. HAL was never included in the helloworld app in the first place (only in callmeback). I'm happy to add it in if we'd like.

File: helloworld/frontend/jest.config.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
```
Comment (marcja) [inline helloworld/frontend/jest.config.js:1]: While is this file needed? Isn't it automagically handled by create-react-app? I was able to run tests with and without code coverage without this file in my own test project.

File: helloworld/frontend/package.json [hunk 11]
```diff
@@ -1,38 +1,32 @@
 {
-  "name": "web",
+  "name": "frontend",
   "version": "0.1.0",
-  "private": true,
+  "scripts": {
+    "dev": "next dev",
+    "build": "next build",
+    "start": "next start",
+    "test": "jest --watch",
```
Comment (marcja) [inline helloworld/frontend/package.json:8]: These are different than the standard create-react-app version, which uses react-scripts (and which I think abstracts a lot of the code/config you've had to add explicitly). What motivated the change?

File: helloworld/frontend/package.json [hunk 12]
```diff
@@ -1,38 +1,32 @@
...
-    "eject": "react-scripts eject"
-  },
-  "eslintConfig": {
-    "extends": "react-app"
+    "semantic-ui-css": "^2.4.1",
+    "semantic-ui-react": "^0.88.2",
+    "style-loader": "^1.2.1",
+    "url-loader": "^4.1.0"
   },
-  "browserslist": {
```
Comment (marcja) [inline helloworld/frontend/package.json:26]: These were added by create-react-app. What's the motivation for removing them?

File: helloworld/README.md [hunk 13]
```diff
@@ -1,20 +1,65 @@
...
 
-## Build & Run
+## Prerequisites
 
-See top level README.md
+  * A local Kubernetes cluster
+    * Mac: [Docker for Mac](https://docs.docker.com/docker-for-mac/install/)
+    * Linux: [minikube](https://minikube.sigs.k8s.io/docs/start/)
+  * [Tilt](https://docs.tilt.dev/install.html)
+    * Mac: May need to update homebrew permissions to have write access to /usr/local/bin directory (prompts with necessary commands will surface on the terminal if so)
```
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/README.md:22]: I suspected so, good to know.

File: helloworld/README.md [hunk 14]
```diff
@@ -1,20 +1,65 @@
...
-* [Official Gradle documentation](https://docs.gradle.org)
-* [Spring Boot Gradle Plugin Reference Guide](https://docs.spring.io/spring-boot/docs/2.2.6.RELEASE/gradle-plugin/reference/html/)
+### Start your Kubernetes environment
 
-## Additional Links
-These additional references should also help you:
+#### Mac
 
-* [Gradle Build Scans – insights for your project's build](https://scans.gradle.com#gradle)
+Enabling Kubernetes in Docker (as in the Tilt instructions) allows this to work. Nothing additional needs to be run.
```
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/README.md:30]: @mscharfstein tried this and it seems that it does this automatically. @mscharfstein, could you confirm?

File: helloworld/backend/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+ENV APP_HOME=/usr/app
+
+# By default, Docker runs apps as a root user in the container. We'll create a
+# separate one just for the app.
+RUN addgroup --system spring && adduser --system spring --ingroup spring
+USER spring:spring
+WORKDIR $APP_HOME
+
+# Copy all the JAR file over
+COPY --from=builder --chown=spring:spring $APP_HOME/build/libs/*.jar app.jar
```
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/backend/Dockerfile:36]: I didn't really go too much into looking into Tilt optimization (since @itssumertime was working on that). I'm inclined to track that as a follow up item since this PR is already quite complex. Thoughts?

File: helloworld/backend/build.gradle [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+  implementation 'org.springframework.boot:spring-boot-starter-data-rest'
+  implementation 'org.springframework.boot:spring-boot-starter-data-mongodb'
+  runtimeOnly 'com.h2database:h2'
```
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/backend/build.gradle:19]: I'm not sure, this was directly from the existing app and I was focusing on parity functionality.

File: helloworld/frontend/README.md [hunk 17]
```diff
@@ -1,68 +1,21 @@
...
-
-### Deployment
-
-This section has moved here: https://facebook.github.io/create-react-app/docs/deployment
-
-### `npm run build` fails to minify
-
-This section has moved here: https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify
+```bash
+npm test
```
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/frontend/README.md:20]: Either one works, I can update this with both.

File: helloworld/frontend/jest.config.js (hunk 10 above)
Comment (tariqajyusuf) [reply helloworld/frontend/jest.config.js:1]: Yes, create-react-app does do this automatically but it has to be manually specified in Next.js.

See [Next.js with-jest example](https://github.com/zeit/next.js/tree/canary/examples/with-jest)

File: helloworld/frontend/package.json (hunk 11 above)
Comment (tariqajyusuf) [reply helloworld/frontend/package.json:8]: The motivation for the change was using some of the server-side pre-rendering functionality with Next.js as you and I discussed last week I believe. This required initializing a new repo with next and moving over our components and code.

File: helloworld/frontend/build.gradle [hunk 18]
```diff
@@ -0,0 +1,33 @@
+plugins {
+  id 'com.github.node-gradle.node' version '2.2.3'
+}
+
+repositories {
+  mavenCentral()
+}
+
+node {
```
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/frontend/build.gradle:9]: It's mainly to abstract away the references to multiple build systems in our Dockerfiles and GitHub Actions. It simplifies a lot of those components to just reference a gradle parent project and have gradle manage loading dependencies.

File: helloworld/frontend/package.json (hunk 12 above)
Comment (tariqajyusuf) [reply helloworld/frontend/package.json:26]: I missed it in the conversion. Re-added.

File: helloworld/backend/Dockerfile (hunk 15 above)
[the remark this answers is not in the document]
Comment (tariqajyusuf) [reply helloworld/backend/Dockerfile:36]: Understood, I figured as much but I'm trying to remove as much a SPOF off myself ;)
